feat(jwt-auth): store JWT in the request context

mikyll opened the pull request:
### Description

The changes introduce a new parameter to [jwt-auth](https://apisix.apache.org/docs/apisix/plugins/jwt-auth/) plugin's route schema: `store_in_ctx`. If set to `true` (default is `false`) the plugin will store the validated JWT object in the request context.

This feature is useful especially for custom plugins. For example, my company is currently using a custom ACL plugin, based on Casbin ([lua-casbin](https://github.com/casbin/lua-casbin)), that parses a JWT to extract permissions.

Therefore, I would say this feature is useful for at least for two reasons:

- since the JWT can be removed from the request attributes (via `hide_credential = true` configuration), this provides another way to pass the token, without exposing it;
- this prevents custom plugins from duplicated code (retrieve and parse JWT object), which `jwt-auth` plugin already does.

Fixes #11281 

### Checklist

- [x] I have explained the need for this PR and the problem it solves
- [x] I have explained the changes or the new features added to this PR
- [x] I have added tests corresponding to this change
- [x] I have updated the documentation to reflect this change
- [x] I have verified that this change is backward compatible (If not, please discuss on the [APISIX mailing list](https://github.com/apache/apisix/tree/master#community) first)

<!--

Note

1. Mark the PR as draft until it's ready to be reviewed.
2. Always add/update tests for any changes unless you have a good reason.
3. Always update the documentation to reflect the changes made in the PR.
4. Make a new commit to resolve conversations instead of `push -f`.
5. To resolve merge conflicts, merge master instead of rebasing.
6. Use "request review" to notify the reviewer after making changes.
7. Only a reviewer can mark a conversation as resolved.

-->

juzhiyuan commented:
@mikyll Thank you! I will request reviewers.

bzp2010 commented on apisix/plugins/jwt-auth.lua:
```diff
@@ -295,6 +299,10 @@ local function find_consumer(conf, ctx)
     end
 
+    if conf.store_in_ctx then
+        ctx.jwt_obj = jwt_obj
```
Awesome, maybe it would be better called `jwt_payload`, or `jwt_auth_payload`? 

Use `payload`, (and specifically prefix it with the jwt-auth plugin name, as some other plugins such as openid-connect also use JWT, which may cause conflicts

bzp2010 reviewed (approved):
LGTM

> One suggestion for the test, it would be nice if it could be done in a clearer order like “Setup 1 - Test 1 - Setup 2 - Test 2”.

mikyll commented:
> LGTM
> 
> > One suggestion for the test, it would be nice if it could be done in a clearer order like “Setup 1 - Test 1 - Setup 2 - Test 2”.

@bzp2010 I moved the tests in a more proper order and fixed the docs a little 🙂
